[Blogging Prompts] Create Prompts List item

thomashorta opened the pull request:
Part of #17125

Add the UI component for the Prompts List item matching UI specs, and use it in the Prompts List screen.

_Note: this is still using mocked data!_

Video Demo:

https://user-images.githubusercontent.com/5091503/210013429-48318b86-b773-4b95-94b1-12358fe4f41d.mp4

To test:
1. Do the steps to enable the Prompts List feature flag found in the test section of #17675
2. Sign in (if not signed in already)
3. Go to the home dashboard (My Site -> Home)
4. **Verify** the Prompts card is shown
5. Tap the 3-dot menu in the Prompts card
6. Tap the "View more prompts" option
7. **Verify** the Prompts List activity is shown
8. **Verify** it shows a "Loading..." text for a while
9. **Verify** the content is updated with a fake list of 11 items, matching the design spec

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
UI changes only, which are not currently covered by tests.

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
create BloggingPromptsListItem component

Commit:
create a Preview Provider for BloggingPromptsListItem

Commit:
use the Prompts List Item component in the list screen

Commit:
add list separators and background color

Commit:
use Serif font for prompts list text to match specs

Commit:
use UiStringPluralRes and string resources in Prompts List item

Commit:
rename string resources and add missing resource

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/promptslist/compose/BloggingPromptsListItem.kt:
```diff
@@ -0,0 +1,118 @@
+    modifier: Modifier = Modifier,
+) {
+    val context = LocalContext.current
+    val dateFormat = remember(context) {
+        SimpleDateFormat(MONTH_DAY_FORMAT, LocaleManager.getSafeLocale(context))
+    }
```
@RenanLukas, I would also like your opinion and thoughts on this specific part.

I was in doubt if this date formatting should be part of the presentation layer or if the `BloggingPromptsListItemModel` should already have the `String` ready for display (meaning this formatting would need to happen in the UseCase, probably).

Also, having the "formatter" instantiated here will make each item in the list have its own instance of it, even though all of them could be the same. I thought about receiving that formatter as an argument, which would hoist its initialization to the `Screen` level, but then this individual component gets a bit harder to Preview (as it will depend on an object that depends on the context and affects the UI directly) and definitely harder to test (which we are not doing now, anyway).

wdyt?

RenanLukas replied:
Hey @thomashorta  👋 

>if this date formatting should be part of the presentation layer or if the BloggingPromptsListItemModel should already have the String ready for display (meaning this formatting would need to happen in the UseCase, probably).
>Also, having the "formatter" instantiated here will make each item in the list have its own instance of it, even though all of them could be the same. I thought about receiving that formatter as an argument, which would hoist its initialization to the Screen level, but then this individual component gets a bit harder to Preview (as it will depend on an object that depends on the context and affects the UI directly) and definitely harder to test (which we are not doing now, anyway).

I usually like to let the View just display the received content (e.g. getting a string instead of directly formatting a `Date` object) for testability, but if we just move this to the `UseCase` level we might harm the reusability considering another screen might require a different date format. IMHO the best place to format the `Date` would be the `ViewModel`, but I think it would be overkill to map the `UseCase` returned object to another object mapped by the `ViewModel`, with all fields being the same except for the formatted `Date`.
How about getting the format (like `MONTH_DAY_FORMAT`) as parameter in the `UseCase`? WDYT? 🤔
```
suspend fun execute(dateFormat: String): Result

//or with a sealed class to make sure the format is valid

suspend fun execute(dateFormat: DateFormatSealedClass): Result
```

If the suggestion above makes sense, maybe we could inject `LocaleManagerWrapper` in the `UseCase` so the static call doesn't harm the testability.

thomashorta replied:
As we talked in DMs I went forward moving this to a `Mapper` class and using the `LocaleManagerWrapper`. So the mapper is injected in the `ViewModel` and Domain -> UI mapping happens there at the presentation layer but not in the Compose / UI scope, which is better for testability, reusability, and performance. Thanks for the review and suggestions!

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/promptslist/compose/BloggingPromptsListItem.kt:
```diff
@@ -0,0 +1,118 @@
+import java.text.SimpleDateFormat
+
+private const val MONTH_DAY_FORMAT = "MMM d"
+private val answeredColor = Color(0xFF008A20)
```
Thinking of reusability/avoiding duplicate code, I was wondering if it would make sense to have this color extracted to a "common" place 🤔
Somewhere in `org.wordpress.android.ui.compose.theme`, maybe? WDYT?

thomashorta replied:
I thought about it and noticed we don't have our own `Color` object. Usually, I like creating an object to hold all static references to colors, do you think this could be useful? Maybe something like:

```
object AppColor {
    @Stable
    val Green50 = Color(0xFF008A20)
    // other colors
}
```

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/promptslist/compose/BloggingPromptsListScreen.kt:
```diff
@@ -30,8 +29,8 @@ import org.wordpress.android.ui.bloggingprompts.promptslist.model.BloggingPrompt
-import org.wordpress.android.util.LocaleManager
-import java.text.DateFormat
+
+private val listBackgroundColor = Color(0xFFE5E5E5)
```
Same question about extracting it to a "common" place.

thomashorta replied:
About this one, I am not so sure, as I don't know how reusable this color is. Also, I found this color on Figma by using a color picker tool on the divider since, for some reason, it just displays as a flat image there, so I don't know if this is a reusable color (with a specific name and everything) or if this is just used here.

I am in favor of putting only reusable colors in an `AppColor` object, but having specific component colors only inside such components. Wdyt?

RenanLukas reviewed (commented):
Thanks for the PR, @thomashorta . It's working as expected 🎉 

There's just one thing I wanted to ask about the app language. Here I had `Portuguese (Brazil)` selected in the App Settings but the formatted date remains in English. Maybe we should be using something else instead of `LocaleManager#getSafeLocale`? 🤔 

<img width="479" alt="image" src="https://user-images.githubusercontent.com/14964993/210462099-0064db74-d93b-43ce-bdd2-1547f63ce325.png">
<img width="479" alt="image" src="https://user-images.githubusercontent.com/14964993/210461800-dc586f4d-cb28-4417-8425-47dc8717d789.png">

Commit:
extract common app colors used in Compose to AppColor object

It contains colors that are part of Automattic's Design System (Color Studio) as
well as colors commonly used in the app (such as DarkGray).

Commit:
remove Prompts List background color to match design specs

Commit:
move date formatting to a Domain <-> UI model mapper

Move the date formatting that was happening in the UI / View layer of the Prompt
List Item to a Mapper and do it in the presentation (VM) layer. This makes the
UseCase more reusable by using a Domain model (from FluxC) and removes the date
formatting objects from Compose UI.

thomashorta commented:
> Thanks for the PR, @thomashorta . It's working as expected 🎉
> 
> There's just one thing I wanted to ask about the app language. Here I had `Portuguese (Brazil)` selected in the App Settings but the formatted date remains in English. Maybe we should be using something else instead of `LocaleManager#getSafeLocale`? 🤔

With the latest changes, using `LocaleManagerWrapper` this seems to be working correctly. Just a note that if the screen is opened already when the language is changed, it might need to be closed and opened again.

Commit:
clean up some imports

Commit:
fix detekt issues

Commit:
make BloggingPromptsListActivity extend LocaleAwareActivity

this fixes the issue of this screen not using the right Locale when the language
settings are changed INSIDE the app.

Commit:
fix detekt issues (unused import)

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/promptslist/compose/BloggingPromptsListScreen.kt:
```diff
@@ -63,26 +58,14 @@ fun BloggingPromptsListScreen(
-            }
+    LazyColumn(Modifier.fillMaxSize()) {
+        itemsIndexed(promptsList) { index, item ->
+            if (index != 0) Divider()
```
💯

RenanLukas reviewed (approved):
Thanks for the changes, @thomashorta 

LGTM :shipit:

thomashorta commented:
> With the latest changes, using `LocaleManagerWrapper` this seems to be working correctly. Just a note that if the screen is opened already when the language is changed, it might need to be closed and opened again.

So, I was wrong about this one, I was under the impression @RenanLukas mentioned the Android System Language Settings, which were working in my last comment. But changing the language inside the Jetpack App settings was still showing the wrong date formatting.

Just have in mind that other strings in that screen are currently hardcoded or just added to `strings.xml`, that's why they are not translated yet.
